src/Shared/RoslynUtils/CodeWriter.cs

```diff
@@ -7,6 +7,8 @@
 
 internal sealed class CodeWriter : IndentedTextWriter
 {
+    public bool ReadFormEmitted { get; set; }
```

Comment:
Is this meant to be here?

Reply:
Yes, I didn't sleep well last night because I was not a fan of what I did here. 😛 

We only want to emit the `ReadFormAsync` call _once_ per handler, regardless of how many form-related parameters exist in the handler. I had a couple of different options here:

- Maintain the state in the `EmitParameterPreparation` method and pass it by ref to each invocation of `EmitFormParameterPreparation`
- Add some `EmittedCode` dictionary to the `CodeWriter` to track the code that was emitted for a given `CodeWriter` instance
- Use a `ReadFormEmitted` property on the `CodeWriter`

The last option is the ickiest and easiest. Opted to maintain the state in the `EmitParameterPreparation` method since the second option is a little too heavy-handed.